adapter/src/ethereum.rs

```diff
@@ -1,101 +1,486 @@
+            ));
+        }
+
+        let validators: Vec<&str> = channel
+            .spec
+            .validators
+            .into_iter()
+            .map(|v| &v.id[..])
+            .collect();
+        let invalid_address_checkum = check_address_checksum(&validators);
+        if invalid_address_checkum {
+            return Err(AdapterError::Configuration(
+                "channel.validators: all addresses are checksummed".to_string(),
+            ));
+        }
+        // check if channel is valid
+        let is_channel_valid = EthereumAdapter::is_channel_valid(&self.config, channel);
+        if is_channel_valid.is_err() {
+            return Err(AdapterError::InvalidChannel(format!(
+                "{}",
+                is_channel_valid.err().unwrap()
+            )));
+        }
+
+        // query the blockchain for the channel status
+        let contract_query = contract.query("states", channel_id, None, Options::default(), None);
+        let channel_status: U256 = contract_query.wait().expect("contract query failed");
+
+        if channel_status != 1.into() {
+            return Err(AdapterError::Configuration(
+                "channel is not Active on the ethereum network".to_string(),
+            ));
+        }
+
         Ok(true)
     }
 
-    fn session_from_token(&self, _token: &str) -> AdapterResult<String> {
-        // @TODO
-        Ok("hello".to_string())
+    fn session_from_token(&mut self, token: &str) -> AdapterResult<Session> {
+        let token_id = token.to_owned()[..16].to_string();
+        let result = self.tokens_verified.get(&token_id);
+        if result.is_some() {
+            return Ok(result.unwrap().to_owned());
+        }
+
+        let verified = match ewt_verify(&token) {
+            Ok(v) => v,
+            Err(e) => return Err(AdapterError::EwtVerifyFailed(format!("{}", e))),
+        };
+
+        let whoami = self.whoami()?;
+        if whoami != verified.from {
+            return Err(AdapterError::Configuration(
+                "token payload.id !== whoami(): token was not intended for us".to_string(),
+            ));
+        }
+
+        let sess = match &verified.payload.identity {
+            Some(identity) => {
+                let (_eloop, transport) =
+                    web3::transports::Http::new(&self.config.ethereum_network).unwrap();
+                let web3 = web3::Web3::new(transport);
+
+                let contract_address =
+                    Address::from_slice(self.config.ethereum_core_address.as_bytes());
+
+                let contract = Contract::from_json(
+                    web3.eth(),
+                    contract_address,
+                    include_bytes!("../contract/Identity.json"),
+                )
+                .unwrap();
+
+                let contract_query =
+                    contract.query("privileges", verified.from, None, Options::default(), None);
+                let priviledge_level: U256 = contract_query.wait().unwrap();
```

Comment:
let's keep this like that for now
but we should change this to an async call ASAP